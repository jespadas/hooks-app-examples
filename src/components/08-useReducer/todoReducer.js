export const todoReducer = (state = [], action) => {

    /*
        The useReducer hook is used for complex state manipulations and state transitions.
        This hook function returns an array with 2 values. The first one is the state value, 
        and the second value is the dispatch function which is further used to trigger an 
        action with the help of array destructuring.
    */

    switch (action.type) {
        case 'add':
            return [...state, action.payload];

        case 'delete':
            return state.filter(todo => todo.id !== action.payload);

        case 'toggle':
            return state.map(todo =>
                (todo.id === action.payload)
                    ? { ...todo, done: !todo.done }
                    : todo
            );

        default:
            return state;
    }

}
