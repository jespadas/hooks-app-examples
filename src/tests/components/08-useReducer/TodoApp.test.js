import { shallow, mount } from 'enzyme';
import { act } from '@testing-library/react';
import { TodoApp } from '../../../components/08-useReducer/TodoApp';
import { demoTodos } from '../../fixtures/demoTodos';

describe('Test on <TodoApp />', () => {

    const wrapper = shallow(<TodoApp />);

    // Mocking localStorage
    Storage.prototype.setItem = jest.fn(() => { });

    test('Should be displayed correctly', () => {

        expect(wrapper).toMatchSnapshot();

    });

    test('Should add a new todo', () => {

        const wrapper = mount(<TodoApp />);

        act(() => {
            wrapper.find('AddForm').prop('handleAddTodo')(demoTodos[0]);
            wrapper.find('AddForm').prop('handleAddTodo')(demoTodos[1]);
        });

        expect(wrapper.find('h1').text().trim()).toBe('TodoApp ( 2 )');

        expect(localStorage.setItem).toHaveBeenCalledTimes(2);

    });

    test('Should delete a todo', () => {

        wrapper.find('AddForm').prop('handleAddTodo')(demoTodos[0]);
        wrapper.find('TodoList').prop('handleDelete')(demoTodos[0].id);

        expect(wrapper.find('h1').text().trim()).toBe('TodoApp ( 0 )');

    });

});