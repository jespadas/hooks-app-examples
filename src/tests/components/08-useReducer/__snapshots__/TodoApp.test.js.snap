// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test on <TodoApp /> Should be displayed correctly 1`] = `
<div>
  <h1>
    TodoApp ( 
    0
     ) 
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-7"
    >
      <TodoList
        handleDelete={[Function]}
        handleToggle={[Function]}
        todos={Array []}
      />
    </div>
    <div
      className="col-5"
    >
      <AddForm
        handleAddTodo={[Function]}
      />
    </div>
  </div>
</div>
`;
